src/stan/services/pathfinder/single.hpp

```diff
@@ -0,0 +1,1104 @@
+      const auto total_size = elbo_draws.cols() + new_draws.cols();
+      constrained_draws_mat
+          = Eigen::Array<double, -1, -1>(names.size(), total_size);
+      Eigen::VectorXd unconstrained_col;
+      Eigen::VectorXd approx_samples_constrained_col;
+      for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
+        unconstrained_col = elbo_draws.col(i);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = elbo_lp_mat.row(i);
+      }
+      for (Eigen::Index i = elbo_draws.cols(), j = 0; i < total_size;
+           ++i, ++j) {
+        unconstrained_col = new_draws.col(j);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = lp_draws.row(j);
+      }
+    } catch (const std::exception& e) {
+      std::string err_msg = e.what();
+      logger.info(path_num + "Final sampling approximation failed with error: "
+                  + err_msg);
+      logger.info(
+          path_num
+          + "Returning the approximate samples used for ELBO calculation: "
+          + err_msg);
+      constrained_draws_mat
+          = Eigen::Array<double, -1, -1>(names.size(), elbo_draws.cols());
+      Eigen::VectorXd approx_samples_constrained_col;
+      Eigen::VectorXd unconstrained_col;
+      for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
+        unconstrained_col = elbo_draws.col(i);
+        constrained_draws_mat.col(i).head(num_unconstrained_params)
+            = constrain_fun(rng, unconstrained_col,
+                            approx_samples_constrained_col);
+        constrained_draws_mat.col(i).tail(2) = elbo_lp_mat.row(i);
+      }
+      lp_ratio = std::move(elbo_best.lp_ratio);
+    }
+  } else {
+    constrained_draws_mat
+        = Eigen::Array<double, -1, -1>(names.size(), elbo_draws.cols());
+    Eigen::VectorXd approx_samples_constrained_col;
+    Eigen::VectorXd unconstrained_col;
+    for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
+      unconstrained_col = elbo_draws.col(i);
+      constrained_draws_mat.col(i).head(num_unconstrained_params)
+          = constrain_fun(rng, unconstrained_col,
+                          approx_samples_constrained_col);
+      constrained_draws_mat.col(i).tail(2) = elbo_lp_mat.row(i);
+    }
+    lp_ratio = std::move(elbo_best.lp_ratio);
+  }
+  parameter_writer(constrained_draws_mat.matrix());
+  parameter_writer();
+  const auto end_optim_time = std::chrono::steady_clock::now();
+  const double optim_delta_time
+      = std::chrono::duration_cast<std::chrono::milliseconds>(
+            end_optim_time - start_optim_time)
+            .count()
+        / 1000.0;
+  const auto time_header = std::string("Elapsed Time: ");
```

Comment:
I don't understand why this one's defined on the outside and other strings on the inside.  Could it be a